src/util/std_expr.cpp

```diff
@@ -43,6 +43,58 @@ exprt conjunction(const exprt::operandst &op)
   }
 }
 
+// Create a lexicographic less-than relation between two tuples of variables.
+// This is used in the implementation of multidimensional decreases clauses.
+exprt create_lexicographic_less_than(
+  const std::vector<symbol_exprt> &small,
+  const std::vector<symbol_exprt> &large)
+{
+  PRECONDITION(small.size() == large.size());
+
+  if(small.size() == 0)
+  {
+    return false_exprt();
+  }
+
+  // Store conjunctions of equalities.
+  // For example, suppose that the two input vectors are <s1, s2, s3> and <l1,
+  // l2, l3>.
+  // Then this vector stores <s1 == l1, s1 == l1 && s2 == l2,
+  // s1 == l1 && s2 == l2 && s3 == l3>.
+  // In fact, the last element is unnecessary, so we do not create it.
+  exprt::operandst consecutive_equalities(small.size());
+
+  consecutive_equalities[0] =
+    binary_relation_exprt(small[0], ID_equal, large[0]);
+  for(unsigned int i = 1; i < consecutive_equalities.size() - 1; i++)
+  {
+    exprt current_component_equality =
+      binary_relation_exprt(small[i], ID_equal, large[i]);
+    consecutive_equalities[i] =
+      and_exprt(consecutive_equalities[i - 1], current_component_equality);
+  }
```

Comment:
I would separate the creation of the equalities and the forming of the conjunction. In particular, we have a `conjunction(...)` helper, which takes `exprt::operandst` as argument.

Reply:
If we use `conjunction(...)` to create conjunctions of equalities of all necessary lengths (i.e. from 1 up to the size of the multidimensional decreases clause), I believe it will take quadratic time in total. Of course, the size of multidimensional decreases clauses is usually small. So the algorithmic performance should not be a concern. Nonetheless, I think it is better to simply traverse the vector and incrementally build conjunctions of equalities, which will take linear time.